torch/_functorch/aot_autograd.py

```diff
@@ -670,17 +670,25 @@ def convert(idx, x):
                 aot_dispatch_base_graph if aot_config.is_export else aot_dispatch_base
             )
 
-        compiler_fn = partial(
-            aot_wrapper_synthetic_base,
-            compiler_fn=compiler_fn,
-            needs_autograd=needs_autograd,
-        )
-        compiler_fn = partial(aot_wrapper_dedupe, compiler_fn=compiler_fn)
-        # You can put more passes here
+        wrappers = [
+            AOTDedupeWrapper(),
+            AOTSyntheticBaseWrapper(trace_joint=needs_autograd),
+            # Add more passes here
+        ]
+        for wrapper in wrappers:
+            flat_fn, fake_flat_args, aot_config, fw_metadata = wrapper.pre_compile(
+                flat_fn, fake_flat_args, aot_config, fw_metadata=fw_metadata
+            )
 
         compiled_fn = compiler_fn(
             flat_fn, fake_flat_args, aot_config, fw_metadata=fw_metadata
         )
+
+        for wrapper in reversed(wrappers):
+            compiled_fn = wrapper.post_compile(
+                compiled_fn, aot_config, fw_metadata=fw_metadata
```

Comment:
In the original code it looks like the aot_config and fw_metadata are based on the values stored as locals in the function but in this version they're based on the last pass to run. Is that intentional?

Maybe they should be stored in the pass class instead of pass into post_compile()?

(I see you did that with AOTSyntheticBaseWrapper but it seems like this API makes it easy to make a mistake when writing passes)

Reply:
This is intentional: unfortunately, passes have dependencies on each other and must be done in a specific order. For example, aot_synthetic_base requires that aot_wrapper_dedup has already been run, as it assumes all the flat_args are deduplicated. 

Now it's possible we can represent this another way, by having compiler wrappers wrap other compiler wrappers, but I'd argue that's honestly just more confusing than having them in a linear order where order matters.

Reply:
> In the original code it looks like the aot_config and fw_metadata are based on the values stored as locals in the function

Correct me if I'm wrong, but originally these are *partial* applications of the lambdas, the actual code looks something like:
```
def synthetic_base_partial(aot_config, fw_metadata): 
   aot_synthetic_base_wrapper(compiler_fn, aot_config, fw_metadata)
  
def dedup_wrapper_partial(aot_config, fw_metadata):
  aot_wrapper_dedup(synthetic_base_partial, aot_config, fw_metadata)
  
 compiled_fn = dedup_wrapper_partial(aot_config, fw_metadata)
```

Then aot_wrapper_dedup looks something like:

```
def aot_wrapper_dedup(compiler_fn, flat_args, aot_config, fw_metadata):
  # pre compile stuff
  new_fw_metadata = # do stuff with metadata 
  # compiler_fn = synthetic_base_partial here
  compiled_fn = compiler_fn(aot_config, new_fw_metadata)
  # post compile stuff 
```

Written this way, you can see that fw_metadata gets edited by aot_wrapper_dedup before being edited by aot_synthetic_base. 

(Well in actuality, aot_synthetic_base is the one that edits fw_metadata in this specific example so technically it doesn't apply to this particular case (since aot_dedup_wrapper doesn't actually edit fw_metadata or aot_config): but aot_synthetic_base's edits *do* affect all the wrappers in aot_dispatch_base and aot_dispatch_autograd, so the same thing applies, just one layer down in the wrapper)

TLDR I think this is the same behavior before and after the diff